### ISSUE TITLE
Unit testing engine/connector contract code

### ISSUE THREAD

**Issue by birchmd** [aurora-is-near/aurora-engine#28]
Presently we can only test contract code via integration tests. Integration tests are obviously important, but I do not think they are a replacement for unit tests. Unit testing encourages good isolation of logic and components so that they can be tested independently. Unit tests are also generally simple to understand, and so can function as a form of documentation for on-boarding new developers to the project (they show how each piece is supposed to work). They also make future modifications to code easier to make because breaking a specific unit test will more clearly show what is broken, as opposed to an integration test which might be more opaque.

The reason unit tests are not possible currently is twofold:

1. Compilation. Some existing tests depend on std, while the contract code is all no-std. This needs to be reconciled.
2. Code organization. Calls to `sdk` functions generally rely on externals which in production are provided to the wasm environment by the Near runtime. In a unit testing environment we should avoid such calls and instead focus on testing pure functions which do not need the context (storage, block height, timestamps, etc) `sdk` calls provide. This may require refactoring some existing code to isolate the pure logic from the `sdk` side-effects.

The purpose of this issue is to resolve the two above items. It would also be nice to write a couple new unit tests as well, but complete unit test coverage should probably be a separate issue, lest the work for this one become too large.

**Comment by birchmd** [comment]
# Proposal for code organization in point 2 of this issue

Define a trait `Sdk` which has the same functions that are currently in `sdk`. We can then have two implementations of `Sdk`: one which delegates to `sdk::exports` functions (side note: is that a good name for that module? The functions are exported from the host, but we are importing them. Maybe something like "external" would be a better name), and one used in testing which has mocked values.

```rust
// sdk.rs

trait Sdk {
    fn read_input(&self, ) -> Vec<u8>;
    fn read_storage(&self, key: &[u8]) -> Option<Vec<u8>>;
    ...
}

struct External;
impl Sdk for External {
    fn read_input(&self, ) -> Vec<u8> {
        // existing implementation using exports
    }
    ...
}

#[cfg(test)]
struct Testing {
    input: Vec<u8>,
    storage_map: HashMap<Vec<u8>, Vec<u8>>,
}
impl Sdk for Testing {
    fn read_input(&self, ) -> Vec<u8> {
        self.input.clone()
    }

    fn read_storage(&self, key: &[u8]) -> Option<Vec<u8>> {
        self.storage_map.get(key).cloned()
    }

    ...
}
```

```rust
// lib.rs

pub extern "C" fn get_nonce() {
    let sdk = External;
    let address = sdk.read_input_arr20();
    let nonce = Engine::get_nonce(&sdk, &Address(address));
    sdk.return_output(&u256_to_arr(&nonce))
}
```

```rust
// engine.rs

    pub fn get_nonce<S: Sdk>(sdk: &S, address: &Address) -> U256 {
        sdk.read_storage(&address_to_key(KeyPrefix::Nonce, address))
            .map(|value| U256::from_big_endian(&value))
            .unwrap_or_else(U256::zero)
    }
```

## Benefits

* All contract code is unit-testable, and tests should be very easy to write/understand
* Makes it clear where host functions are being called (all such places will need to have a type parameter `S: Sdk`)
* Provides a syntactical reminder to try and isolate IO from pure logic (e.g. if you find yourself writing `fn do_some_business_logic<S: Sdk>(sdk: &S) { ... }` then you might consider refactoring to something like `fn do_some_business_logic(arg1, arg2, ...) -> ReturnValue { ... }` to avoid the type parameter

## Drawbacks

* Possible duplication of work done in `near-sdk-sim` (I argue this not really the case because we aren't interested in simulating the full Near runtime, just being able to mock up the few bits we use).
* More boilerplate as all sdk functions need to be implemented twice (I argue this is not a big deal since the sdk functions will rarely change, so after the initial refactoring effort maintenance shouldn't be that bad).


@artob @joshuajbouw what do you think of this? Is it a refactor you think is worth doing during our upcoming tech debt cycle?

**Comment by joshuajbouw** [comment]
Yeah, this is basically what I had in mind when thinking about this. We need to add an `Sdk` trait which then we could mock and use interchangeably.
